content/en/blog/debugging-services-with-istio/index.md

```diff
@@ -0,0 +1,141 @@
+---
+title: "How to debug services with Istio?"
+description: "In this blog, I will explain how Istio can help to solve issues such as encrypting traffic, provide flexible service access control, configure mutual TLS and fine-grained access policies and auditing. "
+date: "2021-03-04T15:15:00"
+author: "[Peter Jausovec](https://peterj.dev)"
+# thumbnail
+thumbnail: "/images/blog/4.jpg"
+tags: ["istio"]
+---
+
+In this blog, we will explain a couple of approaches you can use to debug services running inside your Kubernetes cluster, without deploying the version of the service you want to debug to the cluster.
+
+### Using traffic mirroring
+
+One way to debug the services is to use traffic mirroring. The way traffic mirroring works is that it copies the real traffic and sents it to the mirrored version. Any responses from the mirrored service get dropped, so real traffic is not being affected at all.
+
+The idea is that you have the source code to the service opened in your development IDE. You start the debugger and then, instead of manually interacting with the service or sending the requests to the service running locally, you make a request to the production service. The traffic from that service gets mirrored to the service running locally on your machine.
+
+Since you will be running the service under the debugger on your machine, we can use the service entry resource to point to the service running on your computer, then mirror the traffic to that. For this approach to work, you will have to expose the service running on your machine through a public IP that can be accessible from the cluster. We will then use that IP as an endpoint in the Service entry resource. The diagram below shows for this idea.
+
+![Debugging a local service, exposed through ngrok](debug-local.png)
+
+The traffic or request comes in to through the Hello web and gets mirrored to the external service, defined by the service entry. That service entry is using the static resolution to route to the IP address that's exposing the locally running Greeter service to the public.
+
+Let's deploy the *hello-web.yaml*, *greeter-service-v1.yaml*, and the Virtual service for the Hello web *hello-web-vs.yaml* from the *src/ch7* folder:
```

Comment:
this is referencing `src/ch7` folder -- I think we should either upload these files for readers to download or include the full listings in the blog post.